Dockerfile

```diff
@@ -1,36 +1,43 @@
-FROM registry.access.redhat.com/ubi8/ubi-minimal as builder
-
-# create enviroment variables and files
-ENV ENVFILE=/tmp/envfile
-RUN echo "export ARCH=`uname -m | sed 's/x86_64/amd64/'`" >> ${ENVFILE}
-ENV PATH=$PATH:/usr/local/go/bin
+ARG TARGET_ARCH=amd64
+FROM registry.access.redhat.com/ubi8/ubi-minimal as base
+ARG TARGET_ARCH
 
+# download packages
 RUN microdnf install -y make tar gzip which && microdnf clean all
+RUN export ARCH=$(uname -m | sed 's/x86_64/amd64/'); curl -L https://go.dev/dl/go1.21.6.linux-${ARCH}.tar.gz | tar -C /usr/local -xzf -
+
+
+FROM base as builder
 
-RUN . ${ENVFILE}; curl -L https://go.dev/dl/go1.21.6.linux-${ARCH}.tar.gz | tar -C /usr/local -xzf -
+# in multistage builds the ARG must be renewed
+ARG TARGET_ARCH
 
+# create enviroment variables
+ENV PATH=$PATH:/usr/local/go/bin
+
+# copy the Go Modules manifests and vendor directory
 WORKDIR /workspace
-# Copy the Go Modules manifests and vendor directory
 COPY go.mod go.mod
 COPY go.sum go.sum
 COPY vendor/ vendor/
 
-# Copy the go source
+# copy the go source
 COPY Makefile Makefile
 COPY main.go main.go
 COPY api/ api/
 COPY pkg/ pkg/
 
-# Build
-RUN . ${ENVFILE}; CGO_ENABLED=0 GOOS=linux GOARCH=${ARCH} GO111MODULE=on make build
+# compile for the TARGET_ARCH
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=${TARGET_ARCH} make build
 
 
-FROM registry.access.redhat.com/ubi8/ubi-minimal
+FROM --platform=linux/${TARGET_ARCH} registry.access.redhat.com/ubi8/ubi-micro
```

Comment:
Can we keep `ubi-minimal` here?

Reply:
I was not aware that `ubi-micro` exists. It may be better than `ubi-minimal`, because it is smaller.
We should look into it and maybe use it.

If it is a technical problem to keep `ubi-minimal`, then it's ok to use `ubi-micro`.

Reply:
If we keep `ubi-micro`, could we use it across all build stages?

Reply:
@0xFelix, the same argument as before applies here. The underlying issue here is not about micro versus minimal. It is about how podman manages and utilizes existing images before building a new one.

For instance, consider a scenario where there is a tagged image as `registry.access.redhat.com/ubi8/ubi-micro` from a different architecture in your system, and you have the following Dockerfile:
```
FROM registry.access.redhat.com/ubi8/ubi-micro
RUN uname -m
```
The build of the image will fail because you are trying to execute a binary from an incompatible arch.

Therefore I came with the approach of keeping the builder steps with the host arch, cross compiling and a different image source, while the target images are from the target arch.

Reply:
Thank you for the explanation.

Reply:
I don't quite understand, that means depending on the images I have on my host the build might fail? In this case it only works because `ubi-micro` needs to be pulled when using `ubi-minimal` in the stages before? Or does this only apply to multi-stage builds? From what you say it sounds like a common issue though?

Reply:
Hey @0xFelix, if the `ubi-minimal` image on your system is from your architecture, it should work without any issues. This is because each time we build the image using `ubi-micro`, we specify the architecture using `--platform=linux/${TARGET_ARCH}`.

The example I mentioned 4 days ago was just to explain why I opted for this approach. The main issue here arises because `podman` or `docker` tries to use an existing image for building without checking if the architecture matches. As you mentioned, this is a common problem.

Reply:
I tried it myself, and even though I got it working, for the builder we need to stay on `ubi-minimal`, because `ubi-micro` has no `dnf` to install dependencies of the build. But for the final image we should keep it.